### PR TITLE
Add 3D support to `miss_forest_impute`

### LINKED ISSUE #948: Longitudinal miss_forest_impute
#### This is a function that operates on `.X` or `.layers` so 3D and array-type and encoding is relevant.

**3D**
If 3D allowed:
- [ ] function handles 3D data
- [ ] the test of the function tests both the expected 2D and 3D implementation

If 3D not allowed:
- [ ] only_2D decorator
- [ ] the test of the function tests for failure if 3D is passe

**array-type**
The available array types are np.array, dask.array, and scipy.sparse matrices.
- [ ] function is single-dispatched, with potential not_implemented errors being raised
- [ ] test is parametrized to test the different array-types

**Encoding**
If encoded data is required
- [ ] mention in documentation that encoded data is needed, with ep.pp.encode.

### PR DESCRIPTION
**Pull request by sueoglu**
fixes #948 
Extends `miss_forest_impute` to handle 3D `EHRData` inputs

- Removed `@function_2D_only()` decorator
-  If the input is 3D: flattens the layer to `(n_obs * n_t, n_vars)` (flattening along axis 0) before imputation and reshapes back to `(n_obs, n_vars, n_t)` afterwards using `_apply_over_time_axis`
- Added guard raising `ValueError` when input is 3D but no layer is specified
- Added tests for 3D imputation
- Raising NotImplemented error for unsupported array types (sparse and dask)

Comparison of `miss_forest_impute` and `simple_impute `with mean strategy using physionet2012 dataset with a missingness of 10%: 

```python3
import ehrapy as ep
import ehrdata as ed
import numpy as np

edata_full = ed.dt.physionet2012(layer="tem_data")
edata = edata_full[:200]

full_layer = edata.layers["tem_data"].copy()

rng = np.random.default_rng(42)
non_nan_mask = ~np.isnan(full_layer)
mask_indices = np.where(non_nan_mask)
n_to_mask = int(0.1 * len(mask_indices[0]))
chosen = rng.choice(len(mask_indices[0]), size=n_to_mask, replace=False)

masked_layer = full_layer.copy()
masked_layer[mask_indices[0][chosen], mask_indices[1][chosen], mask_indices[2][chosen]] = np.nan
edata.layers["tem_data"] = masked_layer

edata_mf = edata.copy()
edata_simple = edata.copy()

ep.pp.simple_impute(edata_simple, layer="tem_data", strategy="mean")
ep.pp.miss_forest_impute(edata_mf, layer="tem_data")

true_vals = full_layer[mask_indices[0][chosen], mask_indices[1][chosen], mask_indices[2][chosen]]
mf_vals = edata_mf.layers["tem_data"][mask_indices[0][chosen], mask_indices[1][chosen], mask_indices[2][chosen]]
simple_vals = edata_simple.layers["tem_data"][mask_indices[0][chosen], mask_indices[1][chosen], mask_indices[2][chosen]]

print("True mean:", np.nanmean(true_vals))
print("Miss forest mean:", np.nanmean(mf_vals), "diff:", abs(np.nanmean(mf_vals) - np.nanmean(true_vals)))
print("Simple mean:", np.nanmean(simple_vals), "diff:", abs(np.nanmean(simple_vals) - np.nanmean(true_vals)))

print("True std:", np.nanstd(true_vals))
print("Miss forest std:", np.nanstd(mf_vals), "diff:", abs(np.nanstd(mf_vals) - np.nanstd(true_vals)))
print("Simple std:", np.nanstd(simple_vals), "diff:", abs(np.nanstd(simple_vals) - np.nanstd(true_vals)))

```
Output: 
```
True mean: 70.79201158141458
Miss forest mean: 70.00290934785606 diff: 0.7891022335585234
Simple mean: 71.76743450695217 diff: 0.9754229255375861
True std: 58.31532031908203
Miss forest std: 46.06704701689582 diff: 12.248273302186213
Simple std: 41.852838829183426 diff: 16.462481489898607
```

Note: I used a subset of observations (200) for this evaluation since `miss_forest_impute `got computationally too expensive and crashed (11988 observations × 37 variables × 48 timepoints which after flattening to (n_obs * n_t, n_vars) resulted in a matrix too large for IterativeImputer)

### CODE REVIEW

File: ehrapy/preprocessing/_imputation.py [hunk 1]
```diff
@@ -505,16 +559,11 @@ def miss_forest_impute(
...
     from sklearn.impute import IterativeImputer
 
     try:
-        imp_num = IterativeImputer(
-            estimator=ExtraTreesRegressor(n_estimators=n_estimators, n_jobs=settings.n_jobs),
-            initial_strategy=num_initial_strategy,
-            max_iter=max_iter,
-            random_state=random_state,
-        )
+        # not sure if this should be kept?
```
Comment (eroell) [inline ehrapy/preprocessing/_imputation.py]: this can be removed indeed, you're defining the Imputer in the single-dispatch.

further, there is in Line 568 an unused definition of RandomForestClassifier which you can also throw out.

File: ehrapy/preprocessing/_imputation.py [hunk 2]
```diff
@@ -527,18 +576,26 @@ def miss_forest_impute(
             var_names = edata.var_names
         var_indices = edata.var_names.get_indexer(var_names).tolist()
 
-        if set(var_indices).issubset(_get_non_numerical_column_indices(edata.X)):
-            raise ValueError(
-                "Can only impute numerical data. Try to restrict imputation to certain columns using "
-                "var_names parameter."
-            )
+        mtx = edata.X if layer is None else edata.layers[layer]
+        input_dtype = mtx.dtype if np.issubdtype(mtx.dtype, np.floating) else np.float64
```
Comment (eroell) [inline ehrapy/preprocessing/_imputation.py]: Could you add a quick comment on why this is needed here? :)

File: tests/preprocessing/test_imputation.py [hunk 3]
```diff
@@ -354,10 +354,34 @@ def test_knn_impute_numerical_data(impute_num_edata):
     _base_check_imputation(impute_num_edata, edata_imputed)
 
 
-def test_missforest_impute_3D_edata(edata_blob_small):
-    miss_forest_impute(edata_blob_small, layer="layer_2")
-    with pytest.raises(ValueError, match=r"only supports 2D data"):
-        miss_forest_impute(edata_blob_small, layer=DEFAULT_TEM_LAYER_NAME)
+@pytest.mark.parametrize("edata_mini_3D_missing_values", [True], indirect=True)
```
Comment (eroell) [inline tests/preprocessing/test_imputation.py:357]: Could you add for the "basic" test a parametrization which also checks the array types, where dask raises a valueerror is checked, and it is also checked that this works with sparse (at least in the 2D case then)?

File: ehrapy/preprocessing/_imputation.py [hunk 4]
```diff
@@ -444,8 +444,43 @@ def _knn_impute(
...
+    _raise_array_type_not_implemented(_miss_forest_impute_function, type(arr))
+
+
+@_miss_forest_impute_function.register(sp.coo_array)
+def _(arr: sp.coo_array, num_initial_strategy, n_estimators, max_iter, random_state):
+    _raise_array_type_not_implemented(_miss_forest_impute_function, type(arr))
+
+
+@_miss_forest_impute_function.register(np.ndarray)
+@_miss_forest_impute_function.register(sp.csr_array)
```
Comment (eroell) [inline ehrapy/preprocessing/_imputation.py]: I think it is legit to consider sparse arrays for imputations and make them dense.

Could you mention this in the function docstring?

**Review by eroell — COMMENTED** [review]
Excellent as always!

Next to the comments, a question of interest:
Could you take physionet2012's 3D data, mask ~10% of the values, and impute them with a) missforest b) simple mean impute, and then compare which one is more accurate by comparing the mean of the imputed values with their true mean, and the standard deviation of the imputed values with their true standard deviation?

A code snippet doing so just for the PR comment would be great!

File: tests/preprocessing/test_imputation.py (hunk 3 above)
Comment (sueoglu) [reply tests/preprocessing/test_imputation.py:357]: @eroell alright! but a quick remark: I noticed that the  `_warn_imputation_threshold` warning doesn't work with sparse arrays, even though imputation itself work with them. Should we skip the warning for the sparse arrays so that the imputation works with them or should we raise a NotImplementedError for sparse arrays as well (since they are not supported in` qc_metrics`)

Comment (eroell) [reply tests/preprocessing/test_imputation.py:357]: Separate discussion on warn_imputation_threshold which might be sparse-supported on zulip.

My question here is though, does the imputation here "work" with sparse arrays (as in, keeping them sparse and doing all operations in the sparse representation) or is the imputation just densifying them? :)

Comment (sueoglu) [reply tests/preprocessing/test_imputation.py:357]: it densifies the sparse arrays so they are not fully supported in that sense

Comment (eroell) [reply tests/preprocessing/test_imputation.py:357]: Then I prefer to raise the NotImplementedError; if users are fine with dense, they can go ahead and densify themselves before. If users are not fine with dense, this operation might silently blow up their memory.

File: ehrapy/preprocessing/_imputation.py [hunk 5]
```diff
@@ -444,8 +444,47 @@ def _knn_impute(
             edata.layers[layer][:, imputer_data_indices] = X_imputed
 
 
+@singledispatch
+def _miss_forest_impute_function(arr, num_initial_strategy, n_estimators, max_iter, random_state):
+    _raise_array_type_not_implemented(_miss_forest_impute_function, type(arr))
+
+
+@_miss_forest_impute_function.register(DaskArray)
```
Comment (eroell) [inline ehrapy/preprocessing/_imputation.py]: Not required, as the fallback to line 448 already raises this error

File: ehrapy/preprocessing/_imputation.py [hunk 6]
```diff
@@ -444,8 +444,47 @@ def _knn_impute(
...
+def _(arr: DaskArray, num_initial_strategy, n_estimators, max_iter, random_state):
```
Comment (eroell) [inline ehrapy/preprocessing/_imputation.py]: Not required, as the fallback to line 448 already raises this error

File: ehrapy/preprocessing/_imputation.py [hunk 7]
```diff
@@ -444,8 +444,47 @@ def _knn_impute(
...
+    _raise_array_type_not_implemented(_miss_forest_impute_function, type(arr))
+
+
+@_miss_forest_impute_function.register(sp.coo_array)
```
Comment (eroell) [inline ehrapy/preprocessing/_imputation.py]: Not required, as the fallback to line 448 already raises this error

File: ehrapy/preprocessing/_imputation.py [hunk 8]
```diff
@@ -444,8 +444,47 @@ def _knn_impute(
...
+def _(arr: sp.coo_array, num_initial_strategy, n_estimators, max_iter, random_state):
+    _raise_array_type_not_implemented(_miss_forest_impute_function, type(arr))
+
+
+@_miss_forest_impute_function.register(sp.csr_array)
```
Comment (eroell) [inline ehrapy/preprocessing/_imputation.py]: Not required, as the fallback to line 448 already raises this error

File: ehrapy/preprocessing/_imputation.py [hunk 9]
```diff
@@ -488,13 +530,29 @@ def miss_forest_impute(
     Examples:
         >>> import ehrdata as ed
         >>> import ehrapy as ep
-        >>> edata = ed.dt.mimic_2()
-        >>> edata = ep.pp.encode(edata, autodetect=True)
-        >>> ep.pp.miss_forest_impute(edata)
+        >>> edata_3d = ed.dt.ehrdata_blobs(n_variables=3, n_observations=3, base_timepoints=2, missing_values=0.3)
```
Comment (eroell) [inline ehrapy/preprocessing/_imputation.py]:
```suggestion
        >>> edata = ed.dt.ehrdata_blobs(n_variables=3, n_observations=3, base_timepoints=2, missing_values=0.3)
```

File: ehrapy/preprocessing/_imputation.py [hunk 10]
```diff
@@ -488,13 +530,29 @@ def miss_forest_impute(
...
+        >>> edata_imputed = ep.pp.knn_impute(edata_3d, layer="tem_data", copy=True)
```
Comment (eroell) [inline ehrapy/preprocessing/_imputation.py]:
```suggestion
        >>> edata_imputed = ep.pp.knn_impute(edata, layer="tem_data", copy=True)
```

File: ehrapy/preprocessing/_imputation.py [hunk 11]
```diff
@@ -488,13 +530,29 @@ def miss_forest_impute(
...
+
+        Example Output:
+
+        >>> edata_3d.layers["tem_data"][0, :, :]
```
Comment (eroell) [inline ehrapy/preprocessing/_imputation.py]:
```suggestion
        >>> edata.layers["tem_data"][0, :, :]
```

**Review by eroell — COMMENTED** [review]
Once these comments are addressed, this is ready to merge!
